transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,556 @@
+                        let _ = sender.clone().start_send(event_back);
+                    } else {
+                        log::error!("State mismatch: event for closed connection");
+                    }
+                }
+            }
+
+            // Data needs to be sent on the UDP socket.
+            ToEndpoint::SendUdpPacket(transmit) => self.next_packet_out = Some(transmit),
+            ToEndpoint::Decoupled => self.handle_decoupling()?,
+        }
+        ControlFlow::Continue(())
+    }
+
+    /// Handle datagram received on the socket.
+    /// The datagram content was written into the `socket_recv_buffer`.
+    fn handle_datagram(&mut self, packet: BytesMut, packet_src: SocketAddr) -> ControlFlow<()> {
+        let local_ip = self.channel.socket_addr.ip();
+        // TODO: ECN bits aren't handled
+        let (connec_id, event) =
+            match self
+                .endpoint
+                .handle(Instant::now(), packet_src, Some(local_ip), None, packet)
+            {
+                Some(event) => event,
+                None => return ControlFlow::Continue(()),
+            };
+        match event {
+            quinn_proto::DatagramEvent::ConnectionEvent(event) => {
+                // Redirect the datagram to its connection.
+                if let Some(sender) = self.alive_connections.get_mut(&connec_id) {
+                    match sender.try_send(event) {
+                        Ok(()) => {}
+                        Err(err) if err.is_disconnected() => {
+                            // Connection was dropped by the user.
+                            // Inform the endpoint that this connection is drained.
+                            self.endpoint
+                                .handle_event(connec_id, quinn_proto::EndpointEvent::drained());
+                            self.alive_connections.remove(&connec_id);
+                        }
+                        Err(err) if err.is_full() => {
+                            // Connection is too busy. Drop the datagram to back-pressure the remote.
+                        }
+                        Err(_) => unreachable!("Error is either `Full` or `Disconnected`."),
+                    }
+                } else {
+                    log::error!("State mismatch: event for closed connection");
+                }
+            }
+            quinn_proto::DatagramEvent::NewConnection(connec) => {
+                // A new connection has been received. `connec_id` is a newly-allocated
+                // identifier.
+                debug_assert_eq!(connec.side(), quinn_proto::Side::Server);
+                let connection_tx = match self.new_connection_tx.as_mut() {
+                    Some(tx) => tx,
+                    None => {
+                        log::warn!(
+                            "Endpoint reported a new connection even though server capabilities are disabled."
```

Comment:
I generally prefer to not panic on external code if it can be avoided. Otherwise a bug in `quinn-proto` may cause our nodes to crash.  
Dropping is safe here since nothing changed in our state.

But not a strong opinion. At current point we do know that this code is unreachable [^1], so I am okay wiuth adding a `unreachable!` if you insist :)

[^1]: https://github.com/quinn-rs/quinn/blob/cd2cf756acc9e7774d2c40992e7a355d0fb51b74/quinn-proto/src/endpoint.rs#L251-L258

Reply:
what about a `debug_assert!`?

I agree we don't want to crash in production here since it's fine to keep going.

Reply:
Good idea. Added in 1f83766.